### ISSUE TITLE
Cancelling an async SqlClient operation throws SqlException, not TaskCanceledException

### ISSUE THREAD

**Issue by roji** [dotnet/SqlClient#26]
When using a cancellation token to cancel an SqlClient async operation, an SqlException is thrown rather than TaskCanceledException. This seems like problematic behavior, here's some analysis.

On the problematic side, standard async APIs generally throw TaskCanceledException, and SqlClient is doing something different and therefore unexpected. Also, as a result, canceled SqlClient tasks don't have status Canceled, they're Faulted instead. This makes differentiating cancellations from real exceptions more difficult, and the problem is compounded if the SqlClient call is just part of a larger async operation (which again, will be faulted instead of canceled unless some manual specific handling is involved).

The argument for the current behavior seems to be that all of SqlClient's server and network errors are raised as SqlException, making it easy to catch all database errors under a single exception type. On the other hand, it seems that cancellation isn't really an error condition as it's been requested by the user and is expected to occur. The TaskCanceledException is simply a mechanism for conveying the cancellation up the stack.

For the record, this was originally reported by @mikkeljohnsen for Npgsql in https://github.com/npgsql/npgsql/issues/1146. Whatever is decided here for SqlClient will also be adopted for Npgsql (which currently behaves like SqlClient in this respect).

**Comment by cheenamalhotra** [comment]
Acknowledged.
As it seems to be a breaking API change, we'll consider fixing it with other API changes, when that happens.

**Comment by virzak** [comment]
I ran into this as well.

Having to use a hack to suppress. Is there a better way to handle this in the meantime?

```csharp
catch (OperationCanceledException)
{
}
// Fixme:
// Remove this catch when issue is resolved: https://github.com/dotnet/SqlClient/issues/26
catch (Microsoft.Data.SqlClient.SqlException sqlex)
{
    var errors = sqlex.Errors;
    var isCancelled = false;
    foreach (var error in errors)
    {
        if (!(error is Microsoft.Data.SqlClient.SqlError sqlError))
            continue;
        if (sqlError.Message == "Operation cancelled by user.")
        {
            isCancelled = true;
        }
    }

    if (!isCancelled)
        throw;
}
```

**Comment by proff** [comment]
@virzak, you used cancellationTokenfor cancel operation, so yo can use it for exception check:
```
catch (SqlException ex) when (cancellationToken.IsCancellationRequested && ex.Number == 0 && ex.State == 0 && ex.Class == 11)
{
    throw new OperationCanceledException(cancellationToken);
}
```

**Comment by proff** [comment]
Or for entity framework core
```
services.AddDbContext<DbContext>(
                options => options
                    .UseSqlServer(...)
                    .AddInterceptors(new WrongExceptionOnCancelFixInterceptor())
```
```
    public class WrongExceptionOnCancelFixInterceptor : DbCommandInterceptor
    {
        public override Task CommandFailedAsync(
            DbCommand command,
            CommandErrorEventData eventData,
            CancellationToken cancellationToken = new CancellationToken())
        {
            if (eventData.IsAsync && cancellationToken.IsCancellationRequested
                                  && eventData.Exception is SqlException { Number: 0, State: 0, Class: 11 })
            {
                throw new OperationCanceledException(cancellationToken);
            }

            return base.CommandFailedAsync(command, eventData, cancellationToken);
        }
    }
```

**Comment by Wraith2** [comment]
I've just been looking at cancellation for a bug and this seems like a reasonable time to possibly address this issue. It looks like there's only one place that we create cancel exceptions. Why not introduce an app switch that can optionally wrap the sql exception in a TaskCanceledException?

**Comment by roji** [comment]
@Wraith2 since opening this issue (it's been a while 🤣), I'm now pretty much convinced that the correct behavior here should be to throw an OperationCanceledException rather than SqlException (though the former can definitely wrap the latter). So I guess the question is more if you guys feel like the breaking change make sense. Of course, this could also be taken care of in the next version.

**Comment by cheenamalhotra** [comment]
We attempted this for 2.0 but could not make it in due to some issues. We're on agreement to take this in, and would be working on it again for 3.0. @johnnypham can elaborate more on current progress.

**Comment by Wraith2** [comment]
If @johnnypham is already working on it I'll leave it alone again. I just spotted the code that i thought would need changing on the way past fixing something else and it came to mind.

In theory we can change the code to wrap the sql exception in the type we want and put it on a compatibility switch so it's an opt-in behavioural change. If that is done how do consumers like EF detect that the capability is present in the version of the library that they're using and set the flag?

**Comment by roji** [comment]
> In theory we can change the code to wrap the sql exception in the type we want and put it on a compatibility switch so it's an opt-in behavioural change.

It seems that the consensus is to make SqlClient throw OperationCanceledException by default, rather than when a special opt-in is used (hence the proposal to do it for 3.0 as a breaking change) - though if I'm understanding it wrong we could discuss this further. Though it would certainly be reasonable to have an opt-out that restores the current behavior of throwing an unwrapped SqlException).

> If that is done how do consumers like EF detect that the capability is present in the version of the library that they're using and set the flag?

Anyone who's interested can check the same compatibility switch that SqlClient would check. But I can't really imagine EF doing this and modifying its behavior (EF also doesn't care that much about the cancellation/failure distinction)... Until SqlClient is changed to throw OperationCanceledException, all cancellations will be treated as failures by consuming code.

**Comment by danpaul88** [comment]
Given that the 3.0 release was just tagged, I assume this didn't make the cut in the end? Does this mean a fix for this will have to wait for a 4.0 release now, since it was originally waiting to be included in 3.0 due to being a breaking change to the type of exception thrown?

**Comment by johnnypham** [comment]
I made progress on this but was blocked by an existing issue where commands were not canceled reliably in some scenarios. We have two PRs to address it ([#956](https://github.com/dotnet/SqlClient/pull/956) and [#973](https://github.com/dotnet/SqlClient/pull/973)) which we're looking to review soon.

**Comment by roji** [comment]
@JRahnama @David-Engel as part of https://github.com/dotnet/efcore/issues/19526 I'm trying to identify SqlExceptions which in fact represent cancellations. However, there doesn't seem to be any error code identifying cancellation... The SqlException contains two SqlErrors, the first message of which says "A severe error occurred on the current command.  The results, if any, should be discarded." and the second says "Operation cancelled by user.". The class of both errors is 11, but the Number, State, and LineNumber are all 0. I could look at the message, but as that's localized that wouldn't always work and be brittle.

Am I missing something? What would be the proper way to identify an exception representing cancellation?

**Comment by Wraith2** [comment]
Cancelling is supported in TDS by sending an out of band cancel request to the server, an attention message. The client is then required to continue processing incoming data from the server until it sees a packet attention acknowledge. The spec says that:

> When the ignore and EOM bits are set, the server does not send an attention acknowledgment, but instead returns a table response with a single DONE token that has a status of DONE_ERROR to indicate that the incoming request was ignored. 

I'm not clear about why we might be getting that behaviour in the case of cancellation when a request has been completely sent to the server and we're waiting for a response. The way I read the code I can't see us setting the ignore flag in out of band attention requests. I suspect that SQL server might not entirely be adhering to the spec or the behaviour is more subtle than I'm understanding. This bears further investigation imo.

Anyway. Ultimately we end up in this bit of code in [TdsParser.TryProcessDone](https://github.com/dotnet/SqlClient/blob/c56baaabb641de6dd595203f78a2e93b65ac2317/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs#L2908)
```csharp
            // Surface exception for DONE_ERROR in the case we did not receive an error token
            // in the stream, but an error occurred.  In these cases, we throw a general server error.  The
            // situations where this can occur are: an invalid buffer received from client, login error
            // and the server refused our connection, and the case where we are trying to log in but
            // the server has reached its max connection limit.  Bottom line, we need to throw general
            // error in the cases where we did not receive an error token along with the DONE_ERROR.
            if ((TdsEnums.DONE_ERROR == (TdsEnums.DONE_ERROR & status)) && stateObj.ErrorCount == 0 &&
                  stateObj.HasReceivedError == false && (RunBehavior.Clean != (RunBehavior.Clean & run)))
            {
                stateObj.AddError(new SqlError(0, 0, TdsEnums.MIN_ERROR_CLASS, _server, SQLMessage.SevereError(), "", 0));

                if (null != reader)
                {
                    if (!reader.IsInitialized)
                    {
                        run = RunBehavior.UntilDone;
                    }
                }
            }
```

which is synthesizing a generic error because it sees a DONE_ERROR status but has no error tokens processed. This explains the lack of state and line fields, we don't have them because we're making up the error based on a heuristic. That also means that we wouldn't be overwriting any information if we chose to use a state value which was outside the range that SQL server was able to send, a negative number for example. However, i would want to be convinced that cancellation was the only way for this code to be reached which as usual will require some research from the sql team side.

Tracing through the code I find that we're generating SqlError instances for every info message that is sent. Changing database, changing language, etc. That feels wasteful and I wonder if their materialization could be deferred until they are needed so we could avoid materializing them and the strings that the contain.

**Comment by MisinformedDNA** [comment]
Any updates?

**Comment by roji** [comment]
Note that in some scenarios, ExecuteDbDataReaderAsync apparently throws an InvalidOperationException with the message "Operation cancelled by user" (see https://github.com/dotnet/efcore/issues/29861). This makes it even more difficult for calling code to identify cancellation.

**Comment by MxNbrt** [comment]
@JRahnama and @Wraith2 what is the current status of this issue. Right now, i have to use a workaround like the one proposed here https://github.com/dotnet/efcore/issues/19526#issuecomment-985685079 but it feels so wrong. 

Supporting the correct exception type would be a breaking change but it would conform to the .net library coding guidelines and working with the SqlClient API (through ef core) would be much easier. Thx

**Comment by janpalas** [comment]
> Or for entity framework core
> 
> ```
> services.AddDbContext<DbContext>(
>                 options => options
>                     .UseSqlServer(...)
>                     .AddInterceptors(new WrongExceptionOnCancelFixInterceptor())
> ```
> 
> ```
>     public class WrongExceptionOnCancelFixInterceptor : DbCommandInterceptor
>     {
>         public override Task CommandFailedAsync(
>             DbCommand command,
>             CommandErrorEventData eventData,
>             CancellationToken cancellationToken = new CancellationToken())
>         {
>             if (eventData.IsAsync && cancellationToken.IsCancellationRequested
>                                   && eventData.Exception is SqlException { Number: 0, State: 0, Class: 11 })
>             {
>                 throw new OperationCanceledException(cancellationToken);
>             }
> 
>             return base.CommandFailedAsync(command, eventData, cancellationToken);
>         }
>     }
> ```

This solution was working perfectly until EF Core 7. After upgrading to EF Core 7 the method `CommandFailedAsync` in an interceptor gets never called. Instead interceptor's method `CommandCanceledAsync` gets called and afterwards `SqlException` is thrown sometimes. However, within `CommandCanceledAsync` method you have no option to investigate the exception to apply the fix and throw `OperationCancelledException` instead of `SqlException`...

**Comment by roji** [comment]
@janpalas when EF calls CommandCanceleAsync instead of CommandFailedAsync, that means that it really figured out that it's a cancellation (by checking whether the cancellation token as triggered); so I'm not sure there's a reason for you to "investigate" anything in the interceptor?

**Comment by janpalas** [comment]
>  it really figured out that it's a cancellation (by checking whether the cancellation token as triggered)

@roji sadly just partly. Now `CommandCanceledAsync` is correctly called, however sometimes cancellation gets "unwrapped" into `OperationCancelledException` (which is correct), but sometimes into `SqlException`. I adjusted interceptor's code to prevent `SqlException`s emerging from cancellation:

```C#
public override Task CommandCanceledAsync(DbCommand command, CommandEndEventData eventData, CancellationToken cancellationToken = default)
{
    // 20105 = Microsoft.EntityFrameworkCore.Database.Command.CommandCancelled
    if (cancellationToken.IsCancellationRequested && eventData.EventId == 20105)
        throw new OperationCanceledException(cancellationToken);

    return Task.CompletedTask;
}
```

**Comment by roji** [comment]
@janpalas yep, EF makes no attempt to correct SqlClient's behavior here by throwing a different type of exception; it just distinguishes between failure and cancellation. In that sense, as far as I can tell, the situation isn't any *worse* in EF Core 7.0 compared to what it was in 6.0 - it's just that there's a different event for cancellation.

**Comment by janpalas** [comment]
@roji after thorough investigation it gets more interesting: In case of command cancellation, `CommandCanceledAsync` is called **only** before `SqlException` gets thrown. `CommandCanceledAsync` gets never called when `OperationCancelledException` is thrown, ie. `CommandCanceledAsync` gets hit only in minory cases. 

And yet, it is not 100 % reliable: Sometimes `SqlException` is thrown in case of command cancellation, but `CommandCanceledAsync` is not hit...

**Comment by roji** [comment]
@janpalas as these are EF questions, can you please open an issue on the EF repo with a code sample that shows the problematic behavior?

**Comment by BieleckiLtd** [comment]
> @virzak, you used cancellationToken for cancel operation, so yo can use it for check exception:
> 
> ```
> catch (SqlException ex) when (cancellationToken.IsCancellationRequested && ex.Number == 0 && ex.State == 0 && ex.Class == 11)
> {
>     throw new OperationCanceledException(cancellationToken);
> }
> ```
> 
> The error message can be localized.

In this specific scenario - is it necessary to check the properties of `SqlException` when the operation is cancelled by a `CancellationToken`, or is the following simplified approach sufficient?
```
catch (SqlException) when (cancellationToken.IsCancellationRequested)
{
    throw new OperationCanceledException(cancellationToken);
}
```

**Comment by philipborg** [comment]
> In this specific scenario - is it necessary to check the properties of `SqlException` when the operation is cancelled by a `CancellationToken`, or is the following simplified approach sufficient?
> 
> ```
> catch (SqlException) when (cancellationToken.IsCancellationRequested)
> {
>     throw new OperationCanceledException(cancellationToken);
> }
> ```

That code has a race condition and is thus not sufficient for robustness. If there is another SqlException and then the cancellation token triggers you will not get the appropriate exception for what really happened.

**Comment by default0** [comment]
So, any library code not explicitly referencing Microsoft.Data.SqlClient but rather indirectly interacting with it through some store interface (or, e.g. through a provider-agnostic `DbContext`) can either take a hike and just live with forever having broken cancellation behavior, or take a (for potentially many users incorrect/unnecessary) dependency on the Microsoft.Data.SqlClient package in order to work around its broken cancellation behavior and this has been the state of affairs for 8+ years now?

Are there any plans to make a change here? My current read is that I must simply learn to live with it, tell downstream projects to deal with it or take a dependency that makes no sense in the context of my project - to me, all of those are pretty bad options.

Usually, if there is known broken behavior in a Microsoft library one will find best practices or recommendations to work around it - while some of the approaches here (e.g. interceptors in EF Core) seem fine for downstream projects, they do not seem viable for library projects that want to detect cancellation.

Is there a useful workaround that I am missing that would cover my scenario? Any help would be much appreciated, thanks!

**Comment by aomader** [comment]
Are there any plans on addressing this at some point?

**Comment by 0xced** [comment]
Since the [last comment](https://github.com/dotnet/SqlClient/issues/26#issuecomment-859926529) by a SqlClient team member was more than 4 years ago, I implemented my own workaround based on the discussion in this issue.

I implemented it at the EF Core level with a custom [`IExecutionStrategy`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.storage.iexecutionstrategy). You can find all the details about my workaround here: https://github.com/0xced/efmssql/

**Comment by Timovzl** [comment]
@johnnypham Could you kindly provide a status update since your last [comment](https://github.com/dotnet/SqlClient/issues/26#issuecomment-859926529), and possibly bring this back into view to be finished up?

**Comment by Wraith2** [comment]
I think the team has changed around so johnny may no longer be available to look at it. You probably want to tag the @dotnet/sqlclientdevteam and it may be something that @benrr101 be interested in.

**Comment by matthetherington** [comment]
We're using `UseAzureSql` with `EnableRetryOnFailure` - which adds its own ExecutionStrategy and means the fix in https://github.com/dotnet/SqlClient/issues/26#issuecomment-3410889326 replaces it.

So here's a slightly modified version I put together, with `SqlServerRetryingExecutionStrategy` as the base class instead of `SqlServerExecutionStrategy`:

```csharp
#pragma warning disable EF1001
public class FixSqlClientIssue26ExecutionStrategy(ExecutionStrategyDependencies dependencies)
    : SqlServerRetryingExecutionStrategy(dependencies)
#pragma warning restore EF1001
{
    private const string OperationCanceledByUserMessage = "Operation cancelled by user.";

    public static IExecutionStrategy Create(ExecutionStrategyDependencies dependencies) =>
        new FixSqlClientIssue26ExecutionStrategy(dependencies);

    protected override bool ShouldRetryOn(Exception exception)
    {
        if (exception is SqlException sqlException)
        {
            foreach (SqlError error in sqlException.Errors)
            {
                if (
                    string.Equals(
                        error.Message,
                        OperationCanceledByUserMessage,
                        StringComparison.Ordinal
                    )
                )
                {
                    return false;
                }
            }
        }
        return base.ShouldRetryOn(exception);
    }
}
```

used like:

```csharp
opt.UseAzureSql(
    connectionString,
    opts =>
    {
        // The order is important here. Add the ExecutionStrategy after EnableRetryOnFailure
        opts.EnableRetryOnFailure();
        opts.ExecutionStrategy(FixSqlClientIssue26ExecutionStrategy.Create);
    }
)
```

**Comment by elachlan** [comment]
We specifically catch `OperationCanceledException` so that we can return `Status499ClientClosedRequest` from our API. 

The behavior is inconsistent as we are getting `SqlException` occasionally.
```
Microsoft.Data.SqlClient.SqlException: A severe error occurred on the current command.  The results, if any, should be discarded.
Operation cancelled by user.
at Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__195_0(System.Threading.Tasks.Task`1 result) at offset 8
at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke() at offset 36
at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.<ExecuteReaderAsync>d__18.MoveNext() at offset 748
at Microsoft.EntityFrameworkCore.Query.Internal.AsyncEnumerator.<InitializeReaderAsync>d__21.MoveNext() at offset 265
at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.<ExecuteAsync>d__7`2.MoveNext() at offset 147
at Microsoft.EntityFrameworkCore.Query.Internal.AsyncEnumerator.<MoveNextAsync>d__20.MoveNext() at offset 236
at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.<ToListAsync>d__67`1.MoveNext() at offset 213
```

**Comment by 0xced** [comment]
I finally turned my experiment into a proper NuGet package: [WorkaroundSqlClientIssue26](https://www.nuget.org/packages/WorkaroundSqlClientIssue26)

It's super easy to use, see the README on NuGet or in the [repository](https://github.com/0xced/WorkaroundSqlClientIssue26).

@matthetherington This new package supports another execution strategy, so you can chain `EnableRetryOnFailure()` and `WorkAroundSqlClientIssue26()`.

@elachlan If you use EF Core then you'll be able to catch `OperationCanceledException` with this workaround.

**Comment by 0xced** [comment]
## 🔍 Triage Summary (on-demand re-triage)

| Check      | Result                                                                |
| ---------- | --------------------------------------------------------------------- |
| Issue type | Bug (really, it's a bug)                                              |
| Area       | The abyss of forgotten bugs                                           |
| Duplicates | #1250, https://github.com/dotnet/runtime/issues/46835, related: #2037 |
| Regression | Not a regression, the behaviour has always been wrong                 |

### Analysis

Throwing randomly `InvalidOperationException`, `SqlException` or `TaskCanceledException` but with a wrong [cancellation token](https://github.com/dotnet/SqlClient/issues/2037) upon cancellation has cost thousands (millions?) of engineering hours. See also: [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

### Next Steps

- Maintainers should fix the bug and throw an `OperationCanceledException` with the proper `CancellationToken` when cancellation happens.
- Release notes of the next major version should mention that the bug has been fixed.

> **Note**: This triage summary is written by a fed up user. No AI was used to write this summary.
